File: Classes/Audio Engine/BassEffectDAO.m
//
//  BassEffectDAO.m
//  iSub
//
//  Created by Benjamin Baron on 12/4/11.
//  Copyright (c) 2011 Ben Baron. All rights reserved.
//

#import "BassEffectDAO.h"
#import "SavedSettings.h"
#import "Swift.h"

#define RANGE_OF_EXPONENTS 9
#define MAX_GAIN 6
#define DEFAULT_BANDWIDTH 18

static BASS_DX8_PARAMEQ BASS_DX8_PARAMEQFromPoint(float percentX, float percentY, float bandwidth) {
    BASS_DX8_PARAMEQ p;
    p.fCenter = exp2f((percentX * RANGE_OF_EXPONENTS) + 5);
    p.fGain = (.5 - percentY) * (CGFloat)(MAX_GAIN * 2);;
    p.fBandwidth = bandwidth;
    return p;
}

@implementation BassEffectDAO

#pragma mark Lifecycle

- (instancetype)initWithType:(BassEffectType)effectType
{
	if ((self = [super init]))
	{
		_type = effectType;
		[self setup];
	}
	
	return self;
}

- (id)readPlist:(NSString *)fileName 
{  
	NSData *plistData = nil;  
	NSError *error = nil;
	NSPropertyListFormat format;  
	id plist;  
	
	NSString *localizedPath = [[NSBundle mainBundle] pathForResource:fileName ofType:@"plist"];  
	plistData = [NSData dataWithContentsOfFile:localizedPath];   
	
	plist = [NSPropertyListSerialization propertyListWithData:plistData options:NSPropertyListImmutable format:&format error:&error];
	if (!plist) 
	{  
		NSLog(@"Error reading plist from file '%@', error = '%@'", localizedPath, error.localizedDescription);
	}  
	
	return plist;  
}  

- (void)setup
{
	//DLog(@"default presets: %@", self.defaultPresets);
	//DLog(@"user presets: %@", self.userPresets);
	NSMutableDictionary *presetsDict = [NSMutableDictionary dictionaryWithCapacity:0];
	
	[presetsDict addEntriesFromDictionary:self.defaultPresets];
	NSDictionary *userPresets = self.userPresets;
	if (userPresets)
		[presetsDict addEntriesFromDictionary:userPresets];

	_presets = [[NSDictionary alloc] initWithDictionary:presetsDict];
}

#pragma mark Public DAO Methods

NSInteger presetSort(id preset1, id preset2, void *context)
{
    NSInteger presetId1 = [[preset1 objectForKey:@"presetId"] integerValue];
	NSInteger presetId2 = [[preset2 objectForKey:@"presetId"] integerValue];
	
    if (presetId1 < presetId2)
        return NSOrderedAscending;
    else if (presetId1 > presetId2)
        return NSOrderedDescending;
    else
        return NSOrderedSame;
}

- (NSArray *)presetsArray
{
	NSMutableArray *presetsArray = [NSMutableArray arrayWithCapacity:0];
	for (NSString *key in [self.presets allKeys])
	{
		[presetsArray addObject:[self.presets objectForKey:key]];
	}
	
	return [presetsArray sortedArrayUsingFunction:presetSort context:NULL];
}

- (NSDictionary *)userPresets
{
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    return [[defaults objectForKey:@"BassEffectUserPresets"] objectForKey:[@(self.type) stringValue]] ?: @{};
}

- (NSArray *)userPresetsArray
{
	NSMutableArray *presetsArray = [NSMutableArray arrayWithCapacity:0];
	NSDictionary *userPresets = self.userPresets;
	for (NSString *key in [userPresets allKeys])
	{
		[presetsArray addObject:[userPresets objectForKey:key]];
	}

	return [presetsArray sortedArrayUsingFunction:presetSort context:NULL];
}

- (NSArray *)userPresetsArrayMinusCustom
{
	NSMutableArray *presetsArray = [NSMutableArray arrayWithCapacity:0];
	NSDictionary *userPresets = self.userPresets;
		
	for (NSString *key in [userPresets allKeys])
	{
		if ([[[userPresets objectForKey:key] objectForKey:@"presetId"] intValue] != BassEffectTempCustomPresetId)
			[presetsArray addObject:[userPresets objectForKey:key]];
	}
	
	return [presetsArray count] ? [presetsArray sortedArrayUsingFunction:presetSort context:NULL] : nil;
}

- (NSDictionary *)defaultPresets
{
	// Load default presets
	return [[self readPlist:@"BassEffectDefaultPresets"] objectForKey:[@(self.type) stringValue]];
}

- (NSInteger)userPresetsCount
{
	if (self.userPresets)
		return [[self.userPresets allKeys] count]; 
	
	return 0;
}

- (NSInteger)defaultPresetsCount
{
	if (self.defaultPresets)
		return [[self.defaultPresets allKeys] count];
	
	return 0;
}

- (NSInteger)selectedPresetIndex
{
	return [self.presetsArray indexOfObject:self.selectedPreset];
}

- (NSInteger)selectedPresetId
{
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	return [[[defaults objectForKey:@"BassEffectSelectedPresetId"] objectForKey:[@(self.type) stringValue]] intValue];
}

- (void)setSelectedPresetId:(NSInteger)preset
{
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	NSMutableDictionary *selectedPresetIds = [NSMutableDictionary dictionaryWithCapacity:0];
	if ([defaults objectForKey:@"BassEffectSelectedPresetId"])
		[selectedPresetIds addEntriesFromDictionary:[defaults objectForKey:@"BassEffectSelectedPresetId"]];
	
	[selectedPresetIds setObject:@(preset) forKey:[@((NSInteger)self.type) stringValue]];
	[defaults setObject:selectedPresetIds forKey:@"BassEffectSelectedPresetId"];
	[defaults synchronize];
}

- (NSDictionary *)selectedPreset
{
    return [self.presets objectForKey:[@(self.selectedPresetId) stringValue]];
}

- (NSArray *)selectedPresetValues
{
	return [self.selectedPreset objectForKey:@"values"];
}

#ifdef OSX

static CGPoint CGPointFromString(NSString *string)
{
    // Takes in format like:
    // {1.5,3.3}
    
    // Remove the brackets
    NSString *temp;
    temp = [string stringByReplacingOccurrencesOfString:@"{" withString:@""];
    temp = [string stringByReplacingOccurrencesOfString:@"}" withString:@""];
    
    // Split the two values from the comma
    NSArray *parts = [temp componentsSeparatedByString:@","];
    if (parts.count == 2)
    {
        // Create the CGPoint
        CGPoint point;
        point.x = [parts[0] floatValue];
        point.y = [parts[1] floatValue];
        return point;
    }
    else
    {
        // If we can't parse, return an empty CGPoint
        return CGPointZero;
    }
}

#endif

- (BassEffectValue *)valueForIndex:(NSInteger)valueIndex
{
	if (valueIndex < 0)
		return nil;
	
	if (!self.selectedPresetValues)
		return nil;
	
	if (valueIndex >= [self.selectedPresetValues count])
		return nil;
	
	CGPoint point = CGPointFromString(self.selectedPresetValues[valueIndex]);
    BassEffectValue *value = [[BassEffectValue alloc] initWithType:self.type
                                                          percentX:point.x
                                                          percentY:point.y
                                                         isDefault:[[self.selectedPreset objectForKey:@"isDefault"] boolValue]];
	return value;
}

- (void)selectPresetId:(NSInteger)presetId
{
	self.selectedPresetId = presetId;
		
	if (self.type == BassEffectType_ParametricEQ)
	{
		[BassPlayer.shared.equalizer removeAllEqualizerValues];
		
		for (int i = 0; i < [self.selectedPresetValues count]; i++)
		{
			BassEffectValue *value = [self valueForIndex:i];
			[BassPlayer.shared.equalizer addEqualizerValue:BASS_DX8_PARAMEQFromPoint(value.percentX, value.percentY, DEFAULT_BANDWIDTH)];
		}
		
		if (settingsS.isEqualizerOn)
			[BassPlayer.shared.equalizer toggleEqualizer];
	}
	
	[NSNotificationCenter postOnMainThreadWithName:Notifications_ObjcDeleteMe.bassEffectPresetLoaded object:nil userInfo:nil];
}

- (void)selectPresetAtIndex:(NSInteger)presetIndex
{
	if (presetIndex >= [self.presets count])
		return;
	
	self.selectedPresetId = [[self.presetsArray[presetIndex] objectForKey:@"presetId"] intValue];
	
	[self selectPresetId:self.selectedPresetId];
}

- (void)deleteCustomPresetForId:(NSInteger)presetId
{
	if (presetId == self.selectedPresetId)
		self.selectedPresetId = 0;
	
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	NSDictionary *allUserPresets = [defaults objectForKey:[NSString stringWithFormat:@"BassEffectUserPresets"]];
	NSMutableDictionary *mutableAllUserPresets = [NSMutableDictionary dictionaryWithCapacity:0];
	if (allUserPresets)
		[mutableAllUserPresets addEntriesFromDictionary:allUserPresets];
	
	NSMutableDictionary *mutableUserPresets = [NSMutableDictionary dictionaryWithCapacity:0];
	if (self.userPresets)
		[mutableUserPresets addEntriesFromDictionary:self.userPresets];
	
	[mutableUserPresets removeObjectForKey:[@(presetId) stringValue]];
	[mutableAllUserPresets setObject:mutableUserPresets forKey:[@(self.type) stringValue]];
	[defaults setObject:mutableAllUserPresets forKey:@"BassEffectUserPresets"];
	[defaults synchronize];
	
	[self setup];
}

- (void)deleteCustomPresetForIndex:(NSInteger)presetIndex
{
    if (presetIndex >= self.presets.count)
        return;
    
	NSInteger presetId = [[self.presetsArray[presetIndex] objectForKey:@"presetId"] intValue];
	[self deleteCustomPresetForId:presetId];
}

- (void)deleteTempCustomPreset
{
	[self deleteCustomPresetForId:BassEffectTempCustomPresetId];
}

- (void)saveCustomPreset:(NSArray *)arrayOfPoints name:(NSString *)name presetId:(NSInteger)presetId	
{
	if (!name || !arrayOfPoints)
		return;
	
	self.selectedPresetId = presetId;
	
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	NSDictionary *allUserPresets = [defaults objectForKey:[NSString stringWithFormat:@"BassEffectUserPresets"]];
	NSMutableDictionary *mutableAllUserPresets = [NSMutableDictionary dictionaryWithCapacity:0];
	if (allUserPresets)
		[mutableAllUserPresets addEntriesFromDictionary:allUserPresets];
	
	NSMutableDictionary *mutableUserPresets = [NSMutableDictionary dictionaryWithCapacity:0];
	if (self.userPresets)
		[mutableUserPresets addEntriesFromDictionary:self.userPresets];
	
	// Add new temp custom preset
	NSMutableDictionary *newPresetDict = [NSMutableDictionary dictionaryWithCapacity:0];
	[newPresetDict setObject:@(presetId) forKey:@"presetId"];
	[newPresetDict setObject:name forKey:@"name"];
	[newPresetDict setObject:arrayOfPoints forKey:@"values"];
	[newPresetDict setObject:@NO forKey:@"isDefault"];
	[mutableUserPresets setObject:newPresetDict forKey:[@(presetId) stringValue]];
	[mutableAllUserPresets setObject:mutableUserPresets forKey:[@(self.type) stringValue]];
	[defaults setObject:mutableAllUserPresets forKey:@"BassEffectUserPresets"];
	[defaults synchronize];
	
	[self setup];
}

- (void)saveCustomPreset:(NSArray *)arrayOfPoints name:(NSString *)name
{
	NSInteger presetId = BassEffectUserPresetStartId;
	
	NSArray *userPresetsArrayMinusCustom = self.userPresetsArrayMinusCustom;
	if (userPresetsArrayMinusCustom)
	{
		presetId = [[[userPresetsArrayMinusCustom lastObject] objectForKey:@"presetId"] intValue] + 1;
	}
	
	[self saveCustomPreset:arrayOfPoints name:name presetId:presetId];
}

// Takes array of NSStrings containing CGPoints between 0.0 and 1.0
- (void)saveTempCustomPreset:(NSArray *)arrayOfPoints
{
	[self saveCustomPreset:arrayOfPoints name:@"Custom" presetId:BassEffectTempCustomPresetId];
}

@end

File: Classes/Audio Engine/BassEffectDAO.swift
//
//  BassEffectDAO.swift
//  iSub
//
//  Created by Ben Baron on 2/24/24.
//  Copyright © 2024 Ben Baron. All rights reserved.
//

import Foundation
import CocoaLumberjackSwift
import Resolver

@objc enum BassEffectType: Int {
    case parametricEQ = 1
}

// TODO: Finish testing this implementation and remove all unneeded ObjC interop from the various related classes
final class BassEffectDAO {

    @LazyInjected private var player: BassPlayer
    @LazyInjected private var settings: SavedSettings
    
    static let bassEffectTempCustomPresetId = 1000000
    static let bassEffectUserPresetStartId = 1000
    
    private static let bassEffectSelectedPresetIdKey = "BassEffectSelectedPresetId"
    private static let bassEffectUserPresetsKey = "BassEffectUserPresets"
    
    let type: BassEffectType
    
    private(set) var presets: [BassEffectPreset]
    
    var userPresets: [BassEffectPreset] {
        return presets.filter { $0.presetId >= Self.bassEffectUserPresetStartId }
    }
    
    var userPresetsMinusCustom: [BassEffectPreset] {
        return presets.filter { $0.presetId >= Self.bassEffectUserPresetStartId && $0.presetId != Self.bassEffectTempCustomPresetId }
    }
    
    var selectedPreset: BassEffectPreset? {
        return presets.first { $0.presetId == selectedPresetId }
    }
    
    var selectedPresetId: Int {
        get {
            guard let dict = UserDefaults.standard.object(forKey: Self.bassEffectSelectedPresetIdKey) as? [String: Any],
                  let selectedPresetId = dict["\(type.rawValue)"] as? Int else {
                return -1
            }
            return selectedPresetId
        }
        set {
            guard selectedPresetId != newValue else { return }
            
            var dict = UserDefaults.standard.object(forKey: Self.bassEffectSelectedPresetIdKey) as? [String: Any] ?? [:]
            dict["\(type.rawValue)"] = newValue
            UserDefaults.standard.setValue(dict, forKey: Self.bassEffectSelectedPresetIdKey)
            UserDefaults.standard.synchronize()
        }
    }
    
    var selectedPresetIndex: Int {
        return presets.firstIndex { $0.presetId == selectedPresetId } ?? 0
    }
    
    init(type: BassEffectType) {
        self.type = type
        presets = Self.readAllPresets(type: type)
    }
        
    private static func readDefaultPresets(type: BassEffectType) -> [BassEffectPreset] {
        guard let url = Bundle.main.url(forResource: "BassEffectDefaultPresets", withExtension: "plist") else {
            DDLogError("[BassEffectDAO] Failed to read default presets, the plist file doesn't exist in the main bundle")
            return []
        }
        
        do {
            let data = try Data(contentsOf: url)
            let decoded = try PropertyListDecoder().decode([Int: [Int: BassEffectPreset]].self, from: data)
            if let presets = decoded[type.rawValue] {
                return presets.keys.sorted().map { presets[$0]! }
            }
        } catch {
            DDLogError("[BassEffectDAO] Failed to parse default presets: \(error)")
        }
        
        return []
    }
    
    private static func readUserPresets(type: BassEffectType) -> [BassEffectPreset] {
        guard let dict = UserDefaults.standard.object(forKey: bassEffectUserPresetsKey) as? [String: Any] else { return [] }
        
        do {
            let json = try JSONSerialization.data(withJSONObject: dict)
            let decoded = try JSONDecoder().decode([Int: [Int: BassEffectPreset]].self, from: json)
            if let presets = decoded[type.rawValue] {
                return presets.keys.sorted().map { presets[$0]! }
            }
        } catch {
            DDLogError("[BassEffectDAO] Failed to parse user presets: \(error)")
        }
        
        return []
    }
    
    private static func readAllPresets(type: BassEffectType) -> [BassEffectPreset] {
        return Self.readDefaultPresets(type: type) + Self.readUserPresets(type: type)
    }
    
    func selectCurrentPreset() {
        selectPreset(id: selectedPresetId)
    }
    
    func selectPreset(id: Int) {
        guard let index = presets.firstIndex(where: { $0.presetId == id }) else {
            DDLogError("[BassEffectDAO] Failed to select preset with id \(id) because it doesn't exist")
            return
        }
        
        selectPreset(index: index)
    }
    
    func selectPreset(index: Int) {
        guard index >= 0 && index < presets.count else {
            DDLogError("[BassEffectDAO] Failed to select preset at index \(index) because it doesn't exist")
            return
        }

        let preset = presets[index]
        selectedPresetId = preset.presetId
        
        if type == .parametricEQ {
            player.equalizer.removeAllEqualizerValues()
            
            preset.values.forEach { point in
                let value = BASS_DX8_PARAMEQ(fCenter: exp2f((Float(point.x) * Float(RANGE_OF_EXPONENTS)) + 5),
                                             fBandwidth: Float(DEFAULT_BANDWIDTH),
                                             fGain: Float(0.5 - point.y) * Float(MAX_GAIN * 2))
                player.equalizer.addEqualizerValue(value: value)
            }
            
            if settings.isEqualizerOn {
                player.equalizer.toggleEqualizer()
            }
        }
        
        NotificationCenter.postOnMainThread(name: Notifications.bassEffectPresetLoaded)
    }
    
    func deleteCustomPreset(id: Int) {
        let typeKey = "\(type.rawValue)"
        let idKey = "\(id)"
        
        guard var dict = UserDefaults.standard.object(forKey: Self.bassEffectUserPresetsKey) as? [String: Any],
                var typeDict = dict[typeKey] as? [String: Any], let _ = typeDict.removeValue(forKey: idKey) else {
            DDLogError("[BassEffectDAO] Failed to delete custom preset with id \(id) because it doesn't exist")
            return
        }
        
        if id == selectedPresetId {
            selectedPresetId = 0
        }
        
        dict[typeKey] = typeDict
        UserDefaults.standard.setValue(dict, forKey: Self.bassEffectUserPresetsKey)
        UserDefaults.standard.synchronize()
        
        presets = Self.readAllPresets(type: type)
    }
    
    func deleteCustomPreset(index: Int) {
        guard index >= 0 && index < presets.count else {
            DDLogError("[BassEffectDAO] Failed to delete custom preset at index \(index) because it doesn't exist")
            return
        }
        
        deleteCustomPreset(id: presets[index].presetId)
    }
    
    func deleteTempCustomPreset() {
        deleteCustomPreset(id: Self.bassEffectTempCustomPresetId)
    }
    
    func saveCustomPreset(id: Int, name: String, points: [CGPoint]) {
        selectedPresetId = id
        
        let typeKey = "\(type.rawValue)"
        let idKey = "\(id)"
        
        var dict = UserDefaults.standard.object(forKey: Self.bassEffectUserPresetsKey) as? [String: Any] ?? [:]
        var typeDict = dict[typeKey] as? [String: Any] ?? [:]
        
        let preset: [String: Any] = ["presetId": id,
                                     "name": name,
                                     "values": points.map({ NSCoder.string(for: $0) }),
                                     "isDefault": false]
        typeDict[idKey] = preset
        dict[typeKey] = typeDict
        
        UserDefaults.standard.setValue(dict, forKey: Self.bassEffectUserPresetsKey)
        UserDefaults.standard.synchronize()
        
        presets = Self.readAllPresets(type: type)
    }
    
    func saveCustomPreset(name: String, points: [CGPoint]) {
        var id = Self.bassEffectUserPresetStartId
        if let highestUserPresetId = userPresetsMinusCustom.last?.presetId {
            id = highestUserPresetId + 1
        }
        
        saveCustomPreset(id: id, name: name, points: points)
    }
    
    func saveTempCustomPreset(points: [CGPoint]) {
        saveCustomPreset(id: Self.bassEffectTempCustomPresetId, name: "Custom", points: points)
    }
}

struct BassEffectPreset: Codable {
    let presetId: Int
    let name: String
    let isDefault: Bool
    let values: [CGPoint]
    
    enum CodingKeys: String, CodingKey {
        case presetId
        case name
        case isDefault
        case values
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.presetId = try container.decode(Int.self, forKey: .presetId)
        self.name = try container.decode(String.self, forKey: .name)
        self.isDefault = try container.decode(Bool.self, forKey: .isDefault)
        
        let valueStrings = try container.decode([String].self, forKey: .values)
        self.values = valueStrings.map { NSCoder.cgPoint(for: $0) }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.presetId, forKey: .presetId)
        try container.encode(self.name, forKey: .name)
        try container.encode(self.isDefault, forKey: .isDefault)
        
        let valueStrings = self.values.map { NSCoder.string(for: $0) }
        try container.encode(valueStrings, forKey: .values)
    }
}
